tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs

```diff
@@ -0,0 +1,189 @@
+// <copyright file="TraceTagCollection.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace.Tagging
+{
+    internal class TraceTagCollection
+    {
+        // key1=value1,key2=value2
+        private const char TagPairSeparator = ',';
+        private const char KeyValueSeparator = '=';
+
+        // tags with this prefix are propagated horizontally
+        // (i.e. from upstream services and to downstream services)
+        // using the `x-datadog-tags` header
+        private const string PropagatedTagPrefix = "_dd.p.";
+
+        // for now we only expect one trace-level tag:
+        // "_dd.p.upstream_services"
+        private const int DefaultCapacity = 1;
+
+        // "_dd.p.a=b"
+        public const int MinimumPropagationHeaderLength = 9;
+        public const int MaximumPropagationHeaderLength = 512;
+
+        private static readonly char[] TagPairSeparators = { TagPairSeparator };
+
+        private List<KeyValuePair<string, string>> _tags;
+        private string? _cachedPropagationHeader;
+
+        public TraceTagCollection()
+        {
+            _tags = new List<KeyValuePair<string, string>>(DefaultCapacity);
+        }
+
+        public TraceTagCollection(List<KeyValuePair<string, string>> tags)
+        {
+            _tags = tags;
+        }
+
+        public static TraceTagCollection ParseFromPropagationHeader(string? propagationHeader)
```

Comment:
Good question. I went back and forth on this. In this PR, we will only have one trace tags if we receive `x-datadog-tags` from an upstream service (we don't generate any ourselves yet), so I could make the change you suggest.

But after one of my next PRs, we will _always_ have at least one trace tag because we will start generating `_dd.p.upstream_services` ourselves

Reply:
So the answer to "how often?" is: there will _always_ be trace tags. Not only `_dd.p.upstream_services`, but I also want to move other existing tags that should be "trace-level" 
 (like sampling priority) to use this system.